gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java

```diff
@@ -257,6 +257,21 @@ public T construct() {
     };
   }
 
+  private static boolean hasStringKeyType(Type mapType) {
+    // If mapType is not parameterized, cannot assume that key is String, because if it
+    // is not String and does not implement Comparable it would lead to a ClassCastException
+    if (!(mapType instanceof ParameterizedType)) {
+      return false;
+    }
+
+    Type[] typeArguments = ((ParameterizedType) mapType).getActualTypeArguments();
+    if (typeArguments.length == 0) {
+      return false;
+    }
+    // Consider String and supertypes of it
+    return TypeToken.get(typeArguments[0]).getRawType().isAssignableFrom(String.class);
```

Comment:
Maybe this should also be changed to check the resolved type arguments (in case of subtypes with a different count of type parameters), though the underlying issue is #1708, and `LinkedTreeMap` should only be used if `Map` is requested and not for any subtypes. And in that case there is no need to resolve type arguments.